src/lib/pickles/test/ec-scaling-foreign/test.ml

```diff
@@ -0,0 +1,348 @@
+      add_chain ~range_check two_a_x_squared a_x_squared
+    in
+    (* s_numerator = 3 * a_x * a_x + a *)
+    let s_numerator = add ~range_check three_a_x_squared a in
+    let s_denominator = add_chain ~range_check a_y a_y in
+    let s = exists foreign_typ ~compute:(fun () -> ()) in
+    let s_numerator' = mul ~range_check s_denominator s in
+    assert_equal s_numerator s_numerator' ;
+    let s_squared = mul_chain ~range_check s s in
+    let s_squared_sub_a_x = sub_chain ~range_check s_squared a_x in
+    let x = sub_chain ~range_check s_squared_sub_a_x a_x in
+    let x_sub_a_x = sub_chain ~range_check x a_x in
+    let s_mul_a_sub_a_x = mul_chain ~range_check x_sub_a_x s in
+    let neg_y = add_chain ~range_check s_mul_a_sub_a_x a_y in
+    let y = exists foreign_typ ~compute:(fun () -> ()) in
+    let zero' = add_chain ~range_check neg_y y in
+    assert_equal zero zero' ;
+    (* Run the deferred range checks *)
+    List.iter ~f:actual_range_check !range_checks ;
+    (x, y)
+
+  let pt_typ = Typ.tuple2 foreign_typ foreign_typ
+
+  let circuit a pt scalar_bits =
+    let res, _is_zero, _pt =
+      Array.fold
+        ~init:((zero, zero), Boolean.true_, pt)
+        scalar_bits
+        ~f:(fun (acc, is_zero, pt) bit ->
+          let still_zero = Boolean.(is_zero &&& not bit) in
+          let acc_base = if_ is_zero ~typ:pt_typ ~then_:pt ~else_:acc in
+          let acc_res = ec_add pt acc_base in
+          let acc =
+            if_ bit ~typ:pt_typ
+              ~then_:(if_ is_zero ~typ:pt_typ ~then_:pt ~else_:acc_res)
+              ~else_:acc
+          in
+          (* NB: We do an unnecessary final double. Optimise before using anywhere. *)
+          let double_pt = ec_double a pt in
+          (acc, still_zero, double_pt) )
+    in
+    ignore (res : foreign * foreign)
+end
+
+module Foreign_base = struct
+  type foreign = Field.t * Field.t * Field.t
+
+  let zero = (Field.zero, Field.zero, Field.zero)
+
+  let foreign_typ =
+    Typ.tuple3 Field.typ Field.typ Field.typ
+    |> Typ.transport
+         ~there:(fun _ ->
+           (Field.Constant.zero, Field.Constant.zero, Field.Constant.zero) )
+         ~back:(fun _ -> ())
+
+  let assert_equal (x, y, z) (a, b, c) =
+    Field.Assert.equal x a ; Field.Assert.equal y b ; Field.Assert.equal z c
+
+  let add_zero_constraint vars =
+    assert_
+      { annotation = Some __LOC__
+      ; basic =
+          Kimchi_backend_common.Plonk_constraint_system.Plonk_constraint.T
+            (Zero { vars })
+      }
+
+  let range_check (x, y, z) =
+    (* Requires 4 constraints; inject zeros accordingly. *)
+    add_zero_constraint [| x |] ;
+    add_zero_constraint [| y |] ;
+    add_zero_constraint [| z |] ;
+    add_zero_constraint [||]
+
+  (* Assume that we never need to range_check here, because of the range checks
+     elsewhere.
+  *)
+  let add_chain ~range_check:_ (a, b, c) (d, e, f) =
+    let res = exists foreign_typ ~compute:(fun () -> ()) in
```

Comment:
```suggestion
    let res = exists foreign_typ ~compute:Fn.id in
```

Reply:
The house style for this function is still always explicitly a closure. Please stop suggesting this.